src/modm/math/algorithm/prescaler_counter.hpp

```diff
@@ -0,0 +1,186 @@
+/*
+ * Copyright (c) 2024, Niklas Hauser
+ *
+ * This file is part of the modm project.
+ *
+ * This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/.
+ */
+// ----------------------------------------------------------------------------
+
+#pragma once
+
+#include <algorithm>
+#include <cstdint>
+#include <ranges>
+#include <bit>
+
+namespace modm
+{
+
+/**
+ * Computes the best fitting prescaler and counter value from a list.
+ *
+ * @note For ranges the end is *inclusive*: [begin, end]!
+ * @ingroup	modm_math_algorithm
+ */
+template<typename T>
+class
+GenericPrescalerCounter
+{
+public:
+	struct Result
+	{
+		constexpr Result(T input_frequency, T desired_frequency,
+						 uint32_t index, uint32_t prescaler, uint32_t counter) :
+			index{index}, prescaler{prescaler}, counter{counter},
+			frequency{input_frequency / (prescaler * counter)},
+			input_frequency{input_frequency},
+			desired_frequency{desired_frequency},
+			error{1.f - float(input_frequency) / float(prescaler * counter * desired_frequency)}
+		{}
+		/// Zero-indexed prescaler result
+		uint32_t index;
+		/// Prescaler value
+		uint32_t prescaler;
+		/// Counter value
+		uint32_t counter;
+		/// Generated frequency
+		T frequency;
+		/// Input frequency
+		T input_frequency;
+		/// Desired Frequency
+		T desired_frequency;
+		/// Relative Frequency Error
+		float error;
+	};
+
+public:
+	/// From any forward range.
+	static constexpr Result
+	from_range(T input_frequency, T desired_frequency, uint32_t max_counter,
+			   std::ranges::forward_range auto&& prescalers)
+	{
+		const double desired = double(input_frequency) / desired_frequency;
+		const auto [min, max] = std::minmax_element(
+				std::ranges::begin(prescalers), std::ranges::end(prescalers));
+		// clamp to lowest frequency
+		if (desired < (*min * 1)) {
+			return {input_frequency, desired_frequency,
+					uint32_t(std::distance(std::ranges::begin(prescalers), min)),
+					*min, 1};
+		}
+		// clamp to highest frequency
+		if (desired > (*max * max_counter)) {
+			return {input_frequency, desired_frequency,
+					uint32_t(std::distance(std::ranges::begin(prescalers), max)),
+					*max, max_counter};
+		}
+
+		uint32_t out_prescaler{0}, out_index{0}, out_counter{0};
+		double out_error{std::numeric_limits<double>::infinity()};
+		auto fn_compare = [&](uint32_t idx, uint32_t pre, uint32_t cnt)
+		{
+			const double err = std::abs(desired - pre * cnt);
+			if (err < out_error) {
+				out_error = err;
+				out_prescaler = pre;
+				out_counter = cnt;
+				out_index = idx;
+			}
+		};
+		for (uint32_t idx{0}; const auto prescaler : prescalers)
```

Comment:
Maybe
```suggestion
		for (const auto [idx, prescaler] : enumerate(prescalers))
```
?

Reply:
oh, `std::ranges::enumerate` exists, so we can alias this to `modm::enumerate`.

Reply:
:+1:, but that requires gcc 13. Mac OS hosted and AVR currently have 12. Let's upgrade the compilers for the next release and do it afterwards.

Reply:
We need to find a workaround for the Protothread macro issue to upgrade to GCC13…

Reply:
I don't think this is fixable. The whole construct relies on jumping into a statement expression with a case label. According to the gcc documentation this was never valid but is now properly diagnosed since gcc 13. It just happened to work often enough ...

The only acceptable solution I see is abandoning resumable functions.

Reply:
This doesn't seem to work due to an (for me) incomprehensible compiler error about `unsigned int&` lvalue vs `unsigned int` rvalue… help

Reply:
It's a limitation in `modm::enumerate`. The iterator types you have defined (`linear_range`, etc.) return r-values, here plain integers, not references and the implementation can't deal with that. I'll push a fix tonight.

Reply:
It's fixed now. I've also pushed a fixup for a typo in the Nucleo G070 `module.lb` file.

Reply:
Thank you!